apps/dav/lib/UserMigration/CalendarMigrator.php

```
@@ -0,0 +1,451 @@
+				'vCalendar' => $mergedCalendar,
+			];
+		}
+
+		throw new CalendarMigratorException();
+	}
+
+	/**
+	 * @return array<int, array{name: string, vCalendar: VCalendar}>
+	 *
+	 * @throws CalendarMigratorException
+	 */
+	private function getCalendarExports(IUser $user): array {
+		$principalUri = $this->getPrincipalUri($user);
+
+		return array_values(array_filter(array_map(
+			function (ICalendar $calendar) use ($user) {
+				try {
+					return $this->getCalendarExportData($user, $calendar);
+				} catch (CalendarMigratorException $e) {
+					throw new CalendarMigratorException();
+				} catch (InvalidCalendarException $e) {
+					// Allow this exception as invalid (e.g. deleted) calendars are not to be exported
+				}
+			},
+			$this->calendarManager->getCalendarsForPrincipal($principalUri),
+		)));
+	}
+
+	private function getUniqueCalendarUri(IUser $user, string $initialCalendarUri): string {
+		$principalUri = $this->getPrincipalUri($user);
+		$initialCalendarUri = substr($initialCalendarUri, 0, strlen(CalendarMigrator::MIGRATED_URI_PREFIX)) === CalendarMigrator::MIGRATED_URI_PREFIX
+			? $initialCalendarUri
+			: CalendarMigrator::MIGRATED_URI_PREFIX . $initialCalendarUri;
+
+		$existingCalendarUris = array_map(
+			fn (ICalendar $calendar) => $calendar->getUri(),
+			$this->calendarManager->getCalendarsForPrincipal($principalUri),
+		);
+
+		$calendarUri = $initialCalendarUri;
+		$acc = 1;
+		while (in_array($calendarUri, $existingCalendarUris, true)) {
+			$calendarUri = $initialCalendarUri . "-$acc";
+			++$acc;
+		}
+
+		return $calendarUri;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function export(IUser $user, IExportDestination $exportDestination, OutputInterface $output): void {
+		$output->writeln("Exporting calendars…");
+
+		$userId = $user->getUID();
+
+		try {
+			$calendarExports = $this->getCalendarExports($user);
+		} catch (CalendarMigratorException $e) {
+			$output->writeln("<error>Error exporting <$userId> calendars</error>");
```

Please let the exception throw so that the migration service may chose to abort export.